### PR TITLE
add Webhook handler

### PR DESCRIPTION
**Pull request by soulseen**

### CODE REVIEW

File: pkg/handler/general/general_webhook.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	KubeClientSet  client.Client
+	S2iBuilderName string
+	Namespace      string
+}
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+func NewGithubSink(client client.Client) *Trigger {
```
Comment (runzexia) [inline pkg/handler/general/general_webhook.go]: rename this func

File: pkg/apis/devops/v1alpha1/s2ibuilder_types.go [hunk 2]
```diff
@@ -422,6 +422,12 @@ type S2iConfig struct {
 
 	// Whether output build result to status.
 	OutputBuildResult bool `json:"outputBuildResult,omitempty"`
+
```
Comment (runzexia) [inline pkg/apis/devops/v1alpha1/s2ibuilder_types.go]: I recommend adding a token field here.
Avoiding unauthorized people can trigger builds

File: pkg/handler/builder/types.go [hunk 3]
```diff
@@ -1,10 +1,39 @@
...
+	S2iBuilderName = "builders"
+	letterBytes    = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
 )
 
 type HandlerBuilder struct {
 	Pattern string
 	Func    http.HandlerFunc
 }
+
+func GetParamInPath(path, key string) string {
```
Comment (runzexia) [inline pkg/handler/builder/types.go]: This method has the risk of index out of range.

File: pkg/handler/general/general_webhook.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	return nil
+}
+
+var letterRunes = []rune("0123456789abcdefghijklmnopqrstuvwxyz-")
+
+func RandStringRunes(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letterRunes[rand.Intn(len(letterRunes))]
```
Comment (runzexia) [inline pkg/handler/general/general_webhook.go]: Seed is set during initialization, does repeated calls to this method cause repeated results?

File: pkg/apis/devops/v1alpha1/s2ibuilder_types.go (hunk 2 above)
Comment (soulseen) [reply pkg/apis/devops/v1alpha1/s2ibuilder_types.go]: add secret code to authenticate

File: pkg/handler/server.go [hunk 5]
```diff
@@ -1,21 +1,57 @@
...
+limitations under the License.
+*/
+
 package handler
 
 import (
+	"github.com/kubesphere/s2ioperator/pkg/handler/general"
+	"github.com/kubesphere/s2ioperator/pkg/handler/github"
+	"github.com/kubesphere/s2ioperator/pkg/handler/gitlab"
+	"github.com/prometheus/common/log"
```
Comment (runzexia) [inline pkg/handler/server.go]: use klog ?

File: pkg/handler/general/general_webhook.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		return false, nil
+	}
+}
+
+func (g *Trigger) Action() (err error) {
+	namespaceName := types.NamespacedName{
+		Namespace: g.Namespace}
+	repeat := true
+
+	// really we need check name repeat? This is a very low probability thing.
```
Comment (runzexia) [inline pkg/handler/general/general_webhook.go]: prefer use kubernetes generateName https://kubernetes.io/docs/reference/using-api/api-concepts/#generated-values

Comment (runzexia) [reply pkg/handler/general/general_webhook.go]: s2ir namePre is obtained from s2ib

File: pkg/handler/github/github_hook.go [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+		log.Infof("Can not do any action with event type %s", eventType)
+	}
+
+	return err
+}
+
+func (g *Trigger) actionWithPushEvent(event github.PushEvent) error {
+	revisionId := event.HeadCommit.ID
+	creater := s2irunCreatorPre + *event.HeadCommit.Committer.Name
+	s2irunName := s2irunNamePre + guuid.New().String()[:18]
```
Comment (runzexia) [inline pkg/handler/github/github_hook.go]: use generate Name

**Comment by soulseen** [comment]
change to go-restful

File: go.mod [hunk 8]
```diff
@@ -8,11 +8,16 @@ go 1.12
...
+	github.com/emicklei/go-restful v2.9.6+incompatible
+	github.com/emicklei/go-restful-openapi v1.3.0
 	github.com/go-logr/zapr v0.1.1 // indirect
 	github.com/go-openapi/jsonreference v0.19.3 // indirect
 	github.com/go-openapi/spec v0.19.3
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
 	github.com/golang/groupcache v0.0.0-20181024230925-c65c006176ff // indirect
+	github.com/google/go-github v17.0.0+incompatible
+	github.com/google/go-querystring v1.0.0 // indirect
+	github.com/google/uuid v1.1.1
```
Comment (runzexia) [inline go.mod]: clean dep

File: pkg/handler/github/github_hook.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+	if err != nil {
+		log.Errorf("Failed to get S2IBuilder: %s, in namespace %s, with error: %s", g.S2iBuilderName, g.Namespace, err)
+		return nil, err
+	}
+
+	// Check if the event type is in the allow-list, Now just support push event.
+	if eventType != pushEvent {
+		return nil, fmt.Errorf("not support event type %s", eventType)
+	}
+	if instance.Spec.Config.AllowedEvents != nil {
```
Comment (runzexia) [inline pkg/handler/github/github_hook.go]: We only parse push events, do we still need this field?

Comment (soulseen) [reply pkg/handler/github/github_hook.go]: this field is for extend in future. it also can delete.

**Pull request merged**